### PR TITLE
[SPARK-30330][SQL]Support single quotes json parsing for get_json_object and json_tuple

### PR DESCRIPTION
**Pull request by wenfang6**
### What changes were proposed in this pull request?

I execute some query as` select get_json_object(ytag, '$.y1') AS y1 from t4`; SparkSQL return null but  Hive return correct results.
In my production environment, ytag is a json wrapped by single quotes,as follows
```
{'y1': 'shuma', 'y2': 'shuma:shouji'}
{'y1': 'jiaoyu', 'y2': 'jiaoyu:gaokao'}
{'y1': 'yule', 'y2': 'yule:mingxing'}
```
Then l realized some functions including get_json_object and json_tuple does not support  single quotes json parsing.
So l provide this PR to resolve the question.

### Why are the changes needed?

Enabled for Hive compatibility

### Does this PR introduce any user-facing change?

NO

### How was this patch tested?

NEW TESTS

### CODE REVIEW

**Comment by maropu** [comment]
Thanks for your first contribution, @wenfang6. Could you please read the how-to-contribute notes(https://spark.apache.org/contributing.html), then file jira for this issue?

**Comment by maropu** [comment]
cc: @HyukjinKwon

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 1]
```diff
@@ -103,6 +103,7 @@ private[this] object SharedFactory {
   val jsonFactory = new JsonFactoryBuilder()
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: @HyukjinKwon Do you know why we don't use settings from JSONOptions:
https://github.com/apache/spark/blob/67b644c3d74b0587dd5d498b903383ac4de932fe/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala#L133-L147

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: I think both functions were added mainly to support Hive compatibility, and I think that's why.

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: I am not sure if we should match them or not for now considering the original intention of both functions.

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
@wenfang6 can you follow @maropu's advice at https://github.com/apache/spark/pull/26965#issuecomment-567915347?

**Comment by wenfang6** [comment]
> @wenfang6 can you follow @maropu's advice at [#26965 (comment)](https://github.com/apache/spark/pull/26965#issuecomment-567915347)?
OK, Thank you

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 2]
```diff
@@ -39,6 +39,13 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
       Row("alice", "5"))
   }
 
+  test("function get_json_object - support single quotes") {
+    val df: DataFrame = Seq(("""{'name': 'fang', 'age': 5}""")).toDF("a")
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala]: nit: `val df = Seq(("""{'name': 'fang', 'age': 5}""")).toDF("a")`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 3]
```diff
@@ -103,6 +103,7 @@ private[this] object SharedFactory {
...
     // Enabled for Hive compatibility
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: Can you update this comment, too, like `The two options below enabled for Hive compatibility`?

**Review by maropu — APPROVED** [review]
Leave it to @HyukjinKwon .

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala (hunk 2 above)
Comment (wenfang6) [reply sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala]: In this class of JsonFunctionsSuite , other tests are in the following format,
`val df: DataFrame = tuples.toDF("key", "jstring")`
I want to keep the same format as other tests.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala (hunk 3 above)
Comment (wenfang6) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: l have updated this comment,thank you!

**Review by HyukjinKwon — APPROVED** [review]
Thanks @maropu and @MaxGekk. LGTM too

**Comment by HyukjinKwon** [comment]
Merged to master.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala [hunk 4]
```diff
@@ -101,8 +101,9 @@ private[this] object JsonPathParser extends RegexParsers {
 
 private[this] object SharedFactory {
   val jsonFactory = new JsonFactoryBuilder()
-    // Enabled for Hive compatibility
+    // The two options below enabled for Hive compatibility
     .enable(JsonReadFeature.ALLOW_UNESCAPED_CONTROL_CHARS)
+    .enable(JsonReadFeature.ALLOW_SINGLE_QUOTES)
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:106]: We also have the other Json built-in functions, but only these two functions behave differently from the others. It looks weird especially when Hive compatibility is not important for us.

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:106]: But the main reason why these were added is for Hive compatibility IIRC. It's a minimised change to keep the compatibility. Also, I guess these are all instances for Spark JSON functions corresponding to Hive?

Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:106]: I am afraid that end users expect all our JSON functions have the same behavior. That means, either all the functions support SINGLE QUOTE or none of them supports it. Make the behavior configurable? For reading the external JSON data sources, users can control it https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala#L141-L149 

BTW, will this change cause the query result difference?

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:106]: > I am afraid that end users expect all our JSON functions have the same behavior. That means, either all the functions support SINGLE QUOTE or none of them supports it. Make the behavior configurable? For reading the external JSON data sources, users can control it

If we make it configurable, could we make all the behaivours configurable by using `JSONOptions` here as max suggested in https://github.com/apache/spark/pull/26965/files#r360562916 ? Some users might expect the same behaivours between json functions and json external reading, too.

> BTW, will this change cause the query result difference?

Ah, I think yes; after this PR, some query results have changed from `null` to a correct answer. We need to update the migration guide for 3.0.

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:106]: It's too corner case to document .. who will rely on the behaviour of `null` from a single quote JSON ... Users from Hive might depend on the single quote behaviours but It's pretty unlikely for Spark users to depend on single quoted JSONs.

What about we think about fixing it when it's actually requested from users?

Comment (gatorsmile) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:106]: OK. That sounds OK to me. Let us wait for user requests.

**Pull request closed without merging**